lib/MusicBrainz/Server/Controller/WS/js.pm

```diff
@@ -838,6 +853,132 @@ sub type_info : Chained('root') PathPart('type-info') Args(1) {
+    }
+    if ($args{op} !~ /^[gl]te?$/) {
+        $self->detach_with_error($c, q(malformed 'op' parameter));
+    }
+    unless (is_database_row_id($args{id})) {
+        $self->detach_with_error($c, q(malformed 'id' parameter));
+    }
+
+    my @users = $c->model('Editor')->find_possible_spammers(%args);
+
+    $c->res->content_type('application/json; charset=utf-8');
+    $c->res->body(encode_json({
+        users => [map { $c->unsanitized_editor_json($_) } @users],
+    }));
+}
+
+sub mark_spammer : Chained('root') PathPart('mark-spammer') Args(1) {
+    my ($self, $c, $user_id) = @_;
+
+    $c->forward('/user/cookie_login');
+
+    $c->res->content_type('text/plain; charset=utf-8');
+
+    unless ($c->user_exists) {
+        $c->res->body('you are not logged in');
+        $c->res->status(400);
+        return;
+    }
+
+    unless ($c->user->is_account_admin) {
+        $c->res->body('you are not an account admin');
+        $c->res->status(400);
+        return;
+    }
+
+    unless (is_database_row_id($user_id)) {
+        $c->res->body('malformed user id');
+        $c->res->status(400);
+        return;
+    }
+
+    my $editor = $c->model('Editor')->get_by_id($user_id);
+
+    unless ($editor) {
+        $c->res->body("editor $user_id does not exist");
+        $c->res->status(400);
+        return;
+    }
+
+    if ($editor->deleted) {
+        $c->res->body("editor $user_id is deleted");
+        $c->res->status(400);
+        return;
+    }
+
+    my %privilege_updates = (spammer => 1);
+
+    my $undo_param = $c->req->query_params->{undo};
+    if (defined $undo_param) {
+        unless (is_positive_integer($undo_param)) {
+            $c->res->body('malformed undo parameter');
+            $c->res->status(400);
+            return;
+        }
+        my $undo = $undo_param & (
+            $ADDING_NOTES_DISABLED_FLAG |
+            $EDITING_DISABLED_FLAG |
+            $SPAMMER_FLAG |
+            $VOTING_DISABLED_FLAG
+        );
+        if ($undo != $undo_param) {
+            $c->res->body('unsupported undo flags');
+            $c->res->status(400);
+            return;
+        }
+        if ($undo & $SPAMMER_FLAG) {
+            $privilege_updates{spammer} = 0;
+        }
+        if ($undo & $EDITING_DISABLED_FLAG) {
+            $privilege_updates{editing_disabled} = 0;
+        }
+        if ($undo & $VOTING_DISABLED_FLAG) {
+            $privilege_updates{voting_disabled} = 0;
+        }
+        if ($undo & $ADDING_NOTES_DISABLED_FLAG) {
+            $privilege_updates{adding_notes_disabled} = 0;
+        }
+    }
+
+    if ($privilege_updates{spammer} && $editor->is_spammer) {
+        $c->res->body("editor $user_id is already marked as spammer");
+        $c->res->status(400);
+        return;
+    }
+
+    $c->model('Editor')->update_privileges($editor, \%privilege_updates);
+
+    if ($privilege_updates{spammer}) {
+        $c->forward('/discourse/log_out', [$editor]);
```

Comment:
Can we actually delete their accounts? That's what I usually do by hand...

Reply:
https://github.com/metabrainz/musicbrainz-server/pull/3587#discussion_r2180314394